pkg/admission/mutator/shoot.go

```diff
@@ -53,21 +67,183 @@ func (s *shootMutator) Mutate(_ context.Context, new, old client.Object) error {
+	}
+	shootECSClient, err := s.alicloudClientFactory.NewECSClient(region, string(accessKeyID), string(accessKeySecret))
+	if err != nil {
+		return false, err
+	}
+	if exist, err := shootECSClient.CheckIfImageExists(ctx, imageId); err != nil {
+		return false, err
+	} else if exist {
+		if ownedByAliCloud, err := shootECSClient.CheckIfImageOwnedByAliCloud(imageId); err != nil {
+			return false, err
+		} else if ownedByAliCloud {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+func (s *shootMutator) getImageId(ctx context.Context, imageName string, imageVersion *string, imageRegion string, cloudProfileName string) (string, error) {
+	var (
+		cloudProfile    = &corev1beta1.CloudProfile{}
+		cloudProfileKey = kutil.Key(cloudProfileName)
+	)
+	imageId := ""
+	if err := kutil.LookupObject(ctx, s.virtualGardenclient, s.apiReader, cloudProfileKey, cloudProfile); err != nil {
+		return imageId, err
+	}
+	cloudProfileConfig, err := s.getCloudProfileConfig(cloudProfile)
+	if err != nil {
+		return imageId, err
+	}
+	imageId, err = helper.FindImageForRegionFromCloudProfile(cloudProfileConfig, imageName, *imageVersion, imageRegion)
+	return imageId, err
+}
+func (s *shootMutator) getCloudProfileConfig(cloudProfile *corev1beta1.CloudProfile) (*api.CloudProfileConfig, error) {
+	var cloudProfileConfig *api.CloudProfileConfig = &api.CloudProfileConfig{}
+	if _, _, err := s.decoder.Decode(cloudProfile.Spec.ProviderConfig.Raw, nil, cloudProfileConfig); err != nil {
+		return nil, fmt.Errorf("could not decode providerConfig of cloudProfile for '%s': %w", kutil.ObjectName(cloudProfile), err)
+	}
 
-func (s *shootMutator) mutateShootUpdate(oldShoot, shoot *corev1beta1.Shoot) error {
+	return cloudProfileConfig, nil
+}
+func (s *shootMutator) mutateShootUpdate(ctx context.Context, oldShoot, shoot *corev1beta1.Shoot) error {
+	if !equality.Semantic.DeepEqual(oldShoot.Spec, shoot.Spec) {
+		s.mutateShootUpdateForEncryptedDisk(ctx, oldShoot, shoot)
+	}
 	if !equality.Semantic.DeepEqual(oldShoot.Spec, shoot.Spec) {
 		s.mutateForEncryptedSystemDiskChange(oldShoot, shoot)
 	}
-
 	return nil
 }
+func (s *shootMutator) mutateShootUpdateForEncryptedDisk(ctx context.Context, oldshoot, shoot *corev1beta1.Shoot) {
+	for _, worker := range shoot.Spec.Provider.Workers {
+		oldWorker := getWorkerByName(oldshoot, worker.Name)
+		if oldWorker == nil {
+			logger.Info("Set default value of encrypted disk for newly added worker")
+			s.setDefaultForEncryptedDisk(ctx, shoot, &worker)
+			continue
+		}
+		if worker.Volume != nil && worker.Volume.Encrypted == nil && oldWorker.Volume != nil && oldWorker.Volume.Encrypted != nil {
+			logger.Info("Encrypted disk flag for system disk is not set, keep old value")
+			worker.Volume.Encrypted = oldWorker.Volume.Encrypted
+		}
+		oldDataVolumes := oldWorker.DataVolumes
+		for i := range worker.DataVolumes {
+			dataVolume := &worker.DataVolumes[i]
```

Comment:
```suggestion
		for _, dataVolume := range worker.DataVolumes {
```
Don't need to bother the efficiency of pointers

Reply:
No, It doesn't work if we use the element from list. When iterating, element is copied by value not its reference. Then the value can't be written back to shoot

Reply:
My bad. You need to change the dataVolume.